build/helper.py

```diff
@@ -98,80 +102,170 @@ def normalize_string_type(d):
+class ParamListType(Enum):
+    '''Type of parameter list to return'''
+    API_METHOD = 1
+    '''Used for methods param list for the public API
+
+    'skip_self': False,
+    'skip_session_handle': True,
+    'skip_output_parameters': True,
+    'skip_ivi_dance_size_parameter': True,
+    'session_name': 'vi',
+    'name_to_use': 'python_name',
+    'params_to_use': 'parameters'
+    '''
+    IMPL_METHOD = 2
+    '''Used for methods param list for implementation
+
+    'skip_self': False,
+    'skip_session_handle': False,
+    'skip_output_parameters': False,
+    'skip_ivi_dance_size_parameter': False,
+    'session_name': 'vi',
+    'name_to_use': 'python_name',
+    'params_to_use': 'parameters'
+    '''
+    DISPLAY_METHOD = 3
+    '''Used for methods param list for display (rst)
+
+    'skip_self': True,
+    'skip_session_handle': True,
+    'skip_output_parameters': True,
+    'skip_ivi_dance_size_parameter': True,
+    'session_name': 'vi',
+    'name_to_use': 'python_name',
+    'params_to_use': 'parameters'
+    '''
+    LIBRARY_METHOD = 4
+    '''Used for methods param list when calling library
+
+    'skip_self': True,
+    'skip_session_handle': False,
+    'skip_output_parameters': False,
+    'skip_ivi_dance_size_parameter': False,
+    'session_name': 'vi',
+    'name_to_use': 'python_name',
+    'params_to_use': 'parameters'
+    '''
+    LIBRARY_CALL = 5
+    '''Used for methods param list when calling into the DLL
+
+    'skip_self': True,
+    'skip_session_handle': False,
+    'skip_output_parameters': False,
+    'skip_ivi_dance_size_parameter': False,
+    'session_name': 'vi',
+    'name_to_use': 'library_call_name',
+    'params_to_use': 'parameters'
+    '''
+    LIBRARY_CALL_TYPES = 6
+    '''Used for methods param list types when calling into the DLL
+
+    'skip_self': True,
+    'skip_session_handle': False,
+    'skip_output_parameters': False,
+    'skip_ivi_dance_size_parameter': False,
+    'session_name': 'vi',
+    'name_to_use': 'ctypes_type_library_call',
+    'params_to_use': 'parameters'
     '''
-    snippets = ['self']
-    ivi_dance_size_parameter = find_size_parameter(extract_ivi_dance_parameter(parameters), parameters)
-    for x in parameters:
-        skip = False
-        if x['direction'] == 'out' and skip_output_parameters:
-            skip = True
-        if x == ivi_dance_size_parameter and skip_ivi_dance_size_parameter:
-            skip = True
-        if x['name'] == session_name and skip_session_handle:
-            skip = True
-        if not skip:
-            snippets.append(x['python_name'])
-    return ', '.join(snippets)
-
 
-def get_function_parameters_snippet(parameters, session_name='vi'):
-    '''Returns a string suitable for the parameter list of a method given a list of parameter objects
 
-    If session_name set, skip that parameter
+ParamListTypeDefaults = {}
+ParamListTypeDefaults[ParamListType.API_METHOD] = {
```

Comment:
Some of the things in this dictionary are "filtering/sorting" options, and some are "other" options. I think we should split it up.

And in general, I think we can break up this in two logical steps:

1. filter / reorder parameters
2. render the text snippet 

What do you think?

Reply:
Broken into two loops